### ISSUE TITLE
# Triplet network Image Retrieval : Use of CrossBatchMemory() for set of image triplets (query, positive, negative) and without labels.

### ISSUE THREAD

**Issue by abhinav3** [KevinMusgrave/pytorch-metric-learning#414]
Hi,
I have a use case where I've a supervised training triplet data where for a query image, I've got positive looking image and the negative image is randomly chosen. So each datapoint in the training data set is a image triplet of form : (query, positive, negative)
I'm using the triplet margin ranking loss so that the distance b/w query and positive should be less than query and negative image by a margin.
Now, my question is that can I use `CrossBatchMemory()` module to further improve it to take hard negatives from other training triplets and if yes then what would be the syntax for using it? Because I've not got any labels/targets. I'm just optimising the distances in the embedding space.
My current code looks something like below :

```
dist_query_positive, dist_query_negative= mynet(query_img, positive_img, negative_img) #mynet could be a resnet.
loss =  MarginRankingLoss(dist_query_positive, dist_query_negative, margin=0.2)
loss.backward()
optimizer.step()
```

The current triplet network architecture looks like this image : https://i.ibb.co/SmTWR5F/Screenshot-2022-01-12-at-8-18-17-PM.png

**Comment by KevinMusgrave** [comment]
Is there a way to know which datapoints are negatives? If not, I don't think is possible. For example,

- Triplet 1: (dog, dog, bird)
- Triplet 2: (bird, bird, dog)

Here we know the labels, so we can use the (bird, bird) from triplet 2 as additional negatives for the (dog, dog) in triplet 1. But without labels we wouldn't know, so we might mistakenly include the triplet 2 dog as a negative for (dog, dog) in triplet 1.

**Comment by abhinav3** [comment]
Yes Kevin,
We upfront know which data points from other triplets are negatives and which are positives. It's a completely supervised training dataset.

**Comment by KevinMusgrave** [comment]
In that case, this should work:

```python
from pytorch_metric_learning.losses import TripletMarginLoss, CrossBatchMemory

embedding_size = 128 #change this as necessary
memory_size = 1024
loss_fn = TripletMarginLoss(margin=0.2)
loss_fn = CrossBatchMemory(loss_fn, embedding_size, memory_size)

for (imgs, labels) in dataloader:
    embeddings = model(imgs)
    loss = loss_fn(embeddings, labels)
    loss.backward()
    optimizer.step()
```

**Comment by abhinav3** [comment]
Hi kevin,
Thanks for the sample code and suggestions!

But my `dataloader` just gives me three images (query, positive, negative) and there no labels assigned to them.  Pls note that all 3 images belong to same class/label (dog).  Here query and positive images are manually tagged based on their similarity and negative is randomly chosen. Now how do you suggest me to pass `labels` in the `loss_fn` as suggested in the code.

**Comment by KevinMusgrave** [comment]
Hmm I don't understand. You wrote:

> We upfront know which data points from other triplets are negatives and which are positives. It's a completely supervised training dataset.

If that's the case, then you do have labels. If it's not the case, then you can't use ```CrossBatchMemory```

**Comment by KevinMusgrave** [comment]
If samples from all other triplets can be considered negatives, then you can assign labels like this:

```python

labels = torch.tensor([0, 0, 1])

for triplet in dataloader:
    # shape of embeddings is (3, D)
    embeddings = model(triplet)
    loss = loss_fn(embeddings, labels)
    loss.backward()
    optimizer.step()
    
    # shift labels so that they won't overlap with samples in memory
    labels += 2
```

**Comment by abhinav3** [comment]
Thanks kevin, 
This is what was my doubt exactly and you solved it. As you suggested `labels = torch.tensor([0, 0, 1])` would assign new labels to each triplet  and I could  pass them as argument in the `loss_fn`.

Though I've got another doubt. I've got 10k such training triplets in my `dataloader`, then would assigning those many uniques labels cause any issue for model convergence or any where else while training? If no, then I think  I got answers to my original question and we can close this thread. Thanks a Lot for your time and help! I'll try to run this now.
